File: ModFolder/Loenn/entities/CustomWater.lua
local entity = {}

entity.name = "LuckyHelper/CustomWater"
entity.placements = {
    name = "normal",
    data = {
        width = 8,
        height = 8,
        color = "7DCEFA",
        hasBottom = false,
        disableSurfaceJump = false,
        disableSwimRise = false,
        maxSpeedMultiplierX = 1,
        maxSpeedMultiplierY = 1,
        accelerationMultiplierX = 1,
        accelerationMultiplierY = 1,
        killPlayer = false,
        killPlayerDelay = 0,
        playerFlashTimeBeforeKilled = 2,
        playerFlashColor = "50a4d4",
        playerLoseControl = false,
        playerGravity = 0,
        playerCanJump = false,
        refillExtraJump = false,
        disableRay = false,
    }
}

entity.fieldOrder = {
    "x", "y", 
    "width", "height",
    "color", "disableRay", "hasBottom",
    "maxSpeedMultiplierX", "maxSpeedMultiplierY",
    "accelerationMultiplierX", "accelerationMultiplierY",
    "playerFlashTimeBeforeKilled", "playerFlashColor",
    "killPlayerDelay", "killPlayer",
    "playerGravity", "",
    "", "disableSurfaceJump",
    "disableSwimRise", "playerLoseControl",
    "playerCanJump", "refillExtraJump",
}

entity.fieldInformation = 
{
    color = {
        fieldType = "color"
    },
    width = {
        fieldType = "integer"
    },
    height = {
        fieldType = "integer"
    },
    hasBottom = {
        fieldType = "boolean"
    },
    disableSurfaceJump = {
        fieldType = "boolean"
    },
    disableSwimRise = {
        fieldType = "boolean"
    },
    killPlayer = {
        fieldType = "boolean"
    },
    playerLoseControl = {
        fieldType = "boolean"
    },
    playerCanJump = {
        fieldType = "boolean"
    },
    refillExtraJump = {
        fieldType = "boolean"
    },
    disableRay = {
        fieldType = "boolean"
    },
    playerFlashColor = {
        fieldType = "color"
    }
}
entity.borderColor = {30 / 255, 164 / 255, 231 / 255}
entity.fillColor = {85 / 175, 164 / 255, 220 / 255, 0.6}
return entity

File: Entities/Custom/CustomWater.cs
using Celeste.Mod.Entities;
using ExtendedVariants.Module;
using ExtendedVariants.Variants;
using LuckyHelper.Extensions;
using LuckyHelper.Modules;
using Microsoft.Xna.Framework.Graphics;
using MonoMod.Utils;

namespace LuckyHelper.Entities;

[CustomEntity("LuckyHelper/CustomWater")]
[TrackedAs(typeof(Water))]
[Tracked]
public class CustomWater : Water
{
    private Color color;
    private Color preSurfaceColor;
    private Color preFillColor;
    public bool DisableSurfaceJump;
    public bool DisableSwimRise = true;
    public float MaxSpeedMultiplierX;
    public float MaxSpeedMultiplierY;
    public float AccelerationMultiplierX;
    public float AccelerationMultiplierY;

    public bool KillPlayer = false;
    public float KillPlayerDelay = 0;
    public float PlayerFlashTimeBeforeKilled;
    public Color PlayerFlashColor;

    public bool PlayerLoseControl = true;
    public float PlayerGravity;
    public bool PlayerCanJump;
    public bool RefillExtraJump;
    public bool DisableRay = false;

    // todo: gravity

    public CustomWater(EntityData data, Vector2 offset) : base(data, offset)
    {
        // disableSurfaceJump
        DisableSurfaceJump = data.Bool("disableSurfaceJump");
        DisableSwimRise = data.Bool("disableSwimRise");
        MaxSpeedMultiplierX = data.Float("maxSpeedMultiplierX");
        MaxSpeedMultiplierY = data.Float("maxSpeedMultiplierY");
        AccelerationMultiplierX = data.Float("accelerationMultiplierX");
        AccelerationMultiplierY = data.Float("accelerationMultiplierY");
        KillPlayer = data.Bool("killPlayer");
        KillPlayerDelay = data.Float("killPlayerDelay");
        PlayerFlashTimeBeforeKilled = data.Float("playerFlashTimeBeforeKilled");
        PlayerFlashColor = data.HexColor("playerFlashColor");
        PlayerLoseControl = data.Bool("playerLoseControl");
        PlayerGravity = data.Float("playerGravity");
        PlayerCanJump = data.Bool("playerCanJump");
        RefillExtraJump = data.Bool("refillExtraJump");
        DisableRay = data.Bool("disableRay");

        // color
        color = data.HexColor("color");
        var dd = DynamicData.For(this);
        var surfaces = dd.Get<List<Surface>>("Surfaces");
        foreach (Surface surface in surfaces)
        {
            var d = DynamicData.For(surface);
            var mesh = d.Get<VertexPositionColor[]>("mesh");
            int num1 = (int)(surface.Width / 4.0);
            int num2 = (int)(surface.Width * 0.20000000298023224);
            int surfaceStartIndex = (num1 + num2) * 6;
            for (int i = 0; i < num1 * 6; ++i)
                mesh[i].Color = color * 0.3f;
            for (int i = surfaceStartIndex; i < surfaceStartIndex + num1 * 6; ++i)
                mesh[i].Color = color * 0.8f;
            if (DisableRay)
                surface.Rays.Clear();
        }
    }


    public CustomWater(Vector2 position, bool topSurface, bool bottomSurface, float width, float height) : base(
        position, topSurface, bottomSurface, width, height
    )
    {
    }

    public override void Update()
    {
        Color preRayTopColor = RayTopColor;
        RayTopColor = color * 0.6f;
        base.Update();
        RayTopColor = preRayTopColor;

        Player player = CollideFirst<Player>();
        if (player != null)
        {
            if (CustomWaterModule.KillPlayerElapse >= KillPlayerDelay)
            {
                CustomWaterModule.KillPlayerElapse = 0;
                player.Die(Vector2.Zero);
            }

            if (RefillExtraJump)
            {
                var jumpCount = ExtendedVariantsModule.Instance.VariantHandlers[ExtendedVariantsModule.Variant.JumpCount];
                ((JumpCount)jumpCount).RefillJumpBuffer();
                // typeof(JumpCount).GetMethod("RefillJumpBuffer").Invoke(jumpCount, Type.EmptyTypes);
            }
        }
    }

    public override void Render()
    {
        Color preFillColor = FillColor;
        FillColor = color * 0.3f;
        base.Render();
        FillColor = preFillColor;
        
        Player player = CollideFirst<Player>();
        if (player != null)
        {
            player.Sprite.Color = Color.Black ;
            if (KillPlayerDelay - CustomWaterModule.KillPlayerElapse < PlayerFlashTimeBeforeKilled)
            {
                Logger.Warn("TEst", "awa");
                // player.Sprite.Color = player.flash ? PlayerFlashColor : Color.White;
                player.Sprite.Color = Color.Black ;
            }
        }
    }
}